Source/WebCore/html/parser/HTMLToken.h

```diff
@@ -443,3 +443,11 @@ inline const HTMLToken::Attribute* findAttribute(const HTMLToken::AttributeList&
 }
 
 } // namespace WebCore
+
+namespace WTF {
+
+template<> struct VectorTraits<WebCore::HTMLToken::Attribute> : SimpleClassVectorTraits {
```

Comment:
I understand that those 2 vectors have fairly large inline capacity. However, when we add too much data to those vectors, they'd switch to an external VectorBuffer and presumably memcpy wouldn't work anymore?

Reply:
Using `memcpy` for move moves the pointer and the overwrites the original without calling a destructor. It’s true for types like `std::unique_ptr` for that reason (see the `VectorTraits.h` header). There is a separate flag about using `memcpy` for copying that must remain false for the reasons you say above.

Reply:
I am not sure you can initialize `HTMLToken::Attribute` with memset since at least the capacity fields in the vectors will then be incorrect (when a Vector has an inline capacity it sets m_capacity = inlineCapacity in the default constructor.

Reply:
It does make me wonder if there is a better data structure for the attribute lists I the parser. Maybe just a single UChar buffer with a set pattern:

[number of key / value pairs short][key1 length short][first letter of key1 uchar][second letter of key1 uchar]...[last letter of key1 uchar][value1 length short][first letter of value1 uchar][second letter of value1 uchar]...[last letter of value1 uchar][key2 length short]...

Wrap that in a nice little wrapper that allows appending and forward-only iterating and you got yourself something nice.

Reply:
Yes, thanks @weinig, that’s why the patch doesn’t work.

I’ll look into another solution or maybe just land the traits that improve the performance of moving.